### PR TITLE
Feature #164824532 - Migrate code to scxa analytics v3

### PR DESCRIPTION
**Pull request by monicajianu**
See companion [PR in atlas-web-core](https://github.com/ebi-gene-expression-group/atlas-web-core/pull/5) if you want to run this locally.

- Rewrote metadata retrieval methods to use `scxa-analytics-v3` schema and the new JSON facet builder, with plenty of refactoring along the way
- Removed unused methods and associated tests
- Changed `GeneSearchDao` to use new facet builder
- Re-implemented support for missing metadata values, this time in the `TsnePlotService`
- Slightly tweaked logic to retrieve facet tooltips
- Refactoring and improving testing coverage where necessary

### CODE REVIEW

File: src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java [hunk 1]
```diff
@@ -138,58 +143,53 @@ public GeneSearchDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate,
...
+        SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy> characteristicNameFacet =
+                new SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy>()
+                        .setFacetField(CHARACTERISTIC_NAME)
+                        .addSubFacets(ImmutableList.of(characteristicValueFacet))
+                        .isNestedFacet(true);
+
+        SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy> facets =
+                new SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy>()
+                        .setFacetField(EXPERIMENT_ACCESSION)
+                        .addSubFacets(ImmutableList.of(characteristicNameFacet));
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java]: Funky! :rocket:

File: src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java [hunk 2]
```diff
@@ -138,58 +143,53 @@ public GeneSearchDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate,
         );
     }
 
-    // Returns inferred cell types and organism parts for each experiment accession
-    public Map<String, Map<String, List<String>>>
-    getFacets(List<String> cellIds, SingleCellAnalyticsSchemaField... singleCellAnalyticsSchemaFields) {
-        List<SingleCellAnalyticsSchemaField> subFacetFields = Arrays.asList(singleCellAnalyticsSchemaFields);
+    // Returns all the metadata values for each experiment accession, given a subset of metadata types
+    public Map<String, Map<String, List<String>>> getFacets(List<String> cellIds, String... metadataTypes) {
+        List<String> facetFieldValues = Arrays.asList(metadataTypes);
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java]: Instead of this, I think inlining `ImmutableList.of(metadataTypes)` below in the `queryBuilder` is less distracting.

File: src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java [hunk 3]
```diff
@@ -138,58 +143,53 @@ public GeneSearchDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate,
...
+                extractSimpleOrderedMaps(
                         singleCellAnalyticsCollectionProxy
                                 .query(queryBuilder)
                                 .getResponse()
                                 .findRecursive("facets", EXPERIMENT_ACCESSION.name(), "buckets"));
 
-
-        return resultsByExperiment == null ?
+        return isEmpty(resultsByExperiment) ?
                 emptyMap() :
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java]: Again, if `resultsByExperiment` is empty, so will be the stream and the result of the return. You may want to add a test that checks that specific condition if there is none.

File: src/main/java/uk/ac/ebi/atlas/search/JsonGeneSearchController.java [hunk 4]
```diff
@@ -83,13 +85,13 @@ public String search(@RequestParam MultiValueMap<String, String> requestParams)
...
         } else {
             geneQuery =
                     species.map(
                             _species ->
                                     GeneQuery.create(queryTerm, BioentityPropertyName.getByName(category), _species))
-                           .orElseGet(() -> GeneQuery.create(queryTerm, BioentityPropertyName.getByName(category)));
+                            .orElseGet(() -> GeneQuery.create(queryTerm, BioentityPropertyName.getByName(category)));
 
         }
         Optional<ImmutableSet<String>> geneIds = geneIdSearchService.search(geneQuery);
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/search/JsonGeneSearchController.java]: Java 11 includes [`Optional::isEmpty`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html#isEmpty()), could you change it below?

File: src/test/java/uk/ac/ebi/atlas/tsne/TsnePlotServiceIT.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+        assertThat(
+                subject.fetchTSnePlotWithMetadata(
+                        EXPERIMENT_ACCESSION_WITH_MISSING_INFERRED_CELL_TYPE,
+                        perplexity,
+                        metadataType))
+                .containsKeys(StringUtils.capitalize(MISSING_METADATA_VALUE_PLACEHOLDER));
+    }
+
+}
```
Comment (alfonsomunozpomer) [inline src/test/java/uk/ac/ebi/atlas/tsne/TsnePlotServiceIT.java]: Good addition.

File: src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java (hunk 1 above)
Comment (monicajianu) [reply src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java]: Haha, in a good way? I think it's a little bit verbose and can be quite repetitive, which is why I made a lot of options default.

File: src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java (hunk 3 above)
Comment (monicajianu) [reply src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java]: I think the point was to exit early for efficiency's sake? This is a check [you added](https://github.com/ebi-gene-expression-group/atlas/blame/dev/sc/src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java), but the `resultsByExperiment` couldn't be null I think, which is why I replaced it with `isEmpty`.

File: src/main/java/uk/ac/ebi/atlas/metadata/CellMetadataDao.java [hunk 6]
```diff
@@ -4,159 +4,173 @@
...
-                                      Optional<String> cellId,
-                                      Collection<SingleCellAnalyticsSchemaField> fieldsOfInterest) {
-        if (fieldsOfInterest.isEmpty()) {
+    public Map<String, String> getMetadataValuesForCellId(String experimentAccession,
+                                                   String cellId,
+                                                   Collection<String> factorFields,
+                                                   Collection<String> characteristicFields) {
+        if (factorFields.isEmpty() && characteristicFields.isEmpty()) {
             return ImmutableMap.of();
         }
```
[the remark this answers is not in the document]
Comment (monicajianu) [reply src/main/java/uk/ac/ebi/atlas/metadata/CellMetadataDao.java]: The point is to exit early and not actually do the Solr query if you have no metadata (which can happen for some cells).

File: src/main/java/uk/ac/ebi/atlas/search/FacetType.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    public static FacetType fromName(String name) {
+        for (FacetType facetType : FacetType.values()) {
+            if (facetType.name().equalsIgnoreCase(name)) {
+                return facetType;
+            }
+        }
+        return null;
+    }
+}
```
[the remark this answers is not in the document]
Comment (monicajianu) [reply src/main/java/uk/ac/ebi/atlas/search/FacetType.java]: `JsonGeneSearchController` is such a complex class as it is, with lots of heavy logic. I'd be reluctant to add a nested class to it... But I can do it if you want.

File: src/main/java/uk/ac/ebi/atlas/metadata/CellMetadataDao.java [hunk 8]
```diff
@@ -4,159 +4,173 @@
...
-                        .filter(factor -> !factor.toString().equalsIgnoreCase("single_cell_identifier"))
-                        .map(factor -> SingleCellAnalyticsCollectionProxy.factorAsSchemaField(factor.toString()))
-                        .collect(toList());
-
-        if (queryResult.containsKey(CHARACTERISTIC_INFERRED_CELL_TYPE.name())) {
-            metadataFields.add(CHARACTERISTIC_INFERRED_CELL_TYPE);
-        }
+    // Returns a list of available factor types for a given experiment (and optionally, a cell ID).
+    // Note: it is possible that a cell has missing factors...
+    public Set<String> getFactorTypes(String experimentAccession, Optional<String> cellId) {
```
[the remark this answers is not in the document]
Comment (monicajianu) [reply src/main/java/uk/ac/ebi/atlas/metadata/CellMetadataDao.java]: I agree completely, I wasn't a bit fan of the `Optional` in the first place (and IntelliJ told me off about it too). I'll refactor this.

File: src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java (hunk 1 above)
Comment (monicajianu) [reply src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java]: I'm a convert! Helps a little with the verbosity.

Screenshot 2019-05-13 at 15 23 38

File: src/main/java/uk/ac/ebi/atlas/search/FacetType.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (alfonsomunozpomer) [reply src/main/java/uk/ac/ebi/atlas/search/FacetType.java]: Ok, I understand.

File: src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java (hunk 1 above)
Comment (alfonsomunozpomer) [reply src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java]: Funky in a good way. Also, welcome to the dark side. :wink:

File: src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java (hunk 3 above)
Comment (alfonsomunozpomer) [reply src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java]: If I added this check it wasn’t my best moment. :sweat_smile: 
Notice that if `resultsByExperiment` is empty, so will be the resulting stream and the lambda inside the collector won’t be executed.
